File: src/services/EconomyService.cs
using StarSmugglerGo.Domain;
using System;
using System.Collections.Generic;
using System.Linq;

namespace StarSmugglerGo.Services;

public sealed class EconomyService
{
    public Dictionary<string, MarketSnapshot> CreateInitialMarkets(DataSnapshot data, Random random)
    {
        var result = new Dictionary<string, MarketSnapshot>(StringComparer.Ordinal);

        foreach (PortDefinition port in data.Ports)
        {
            result[port.Id] = CreateMarketForPort(port, data.Items, random);
        }

        return result;
    }

    public MarketSnapshot CreateMarketForPort(PortDefinition port, IReadOnlyList<ItemDefinition> items, Random random)
    {
        List<ItemDefinition> availableItems = BuildAvailableItemsForPort(items, port.Zone, random);
        var priceMap = new Dictionary<string, int>(StringComparer.Ordinal);

        foreach (ItemDefinition item in items)
        {
            priceMap[item.Id] = CalculatePrice(item, port.Zone, random);
        }

        return new MarketSnapshot
        {
            PortId = port.Id,
            AvailableItemIds = availableItems.Select(item => item.Id).ToList(),
            PricesByItemId = priceMap,
        };
    }

    public void RefreshAvailableGoods(RunState run, DataSnapshot data, string portId, Random random)
    {
        if (!data.PortsById.TryGetValue(portId, out PortDefinition? port))
        {
            return;
        }

        if (!run.MarketsByPortId.TryGetValue(portId, out MarketSnapshot? market))
        {
            run.MarketsByPortId[portId] = CreateMarketForPort(port, data.Items, random);
            return;
        }

        List<ItemDefinition> availableItems = BuildAvailableItemsForPort(data.Items, port.Zone, random);
        market.AvailableItemIds.Clear();
        market.AvailableItemIds.AddRange(availableItems.Select(item => item.Id));
    }

    public void RefreshAllPrices(RunState run, DataSnapshot data, Random random)
    {
        run.MarketsByPortId.Clear();

        foreach (KeyValuePair<string, MarketSnapshot> pair in CreateInitialMarkets(data, random))
        {
            run.MarketsByPortId[pair.Key] = pair.Value;
        }

        run.JumpsSinceLastUpdate = 0;
    }

    public MarketSnapshot? GetCurrentMarket(RunState run)
    {
        return run.MarketsByPortId.TryGetValue(run.Player.CurrentPortId, out MarketSnapshot? market)
            ? market
            : null;
    }

    public int GetCargoLoad(RunState run)
    {
        return run.Cargo.TotalUnits;
    }

    public IReadOnlyList<ItemDefinition> GetAvailableGoodsForCurrentPort(RunState run, DataSnapshot data)
    {
        MarketSnapshot? market = GetCurrentMarket(run);
        if (market is null)
        {
            return Array.Empty<ItemDefinition>();
        }

        List<ItemDefinition> goods = market.AvailableItemIds
            .Select(itemId => data.ItemsById.TryGetValue(itemId, out ItemDefinition? item) ? item : null)
            .Where(item => item is not null)
            .Cast<ItemDefinition>()
            .ToList();

        return goods;
    }

    public int GetSellableCargoValueAtCurrentPort(RunState run, DataSnapshot data)
    {
        MarketSnapshot? market = GetCurrentMarket(run);
        if (market is null)
        {
            return 0;
        }

        int total = 0;

        foreach ((string itemId, int quantity) in run.Cargo.ItemQuantities)
        {
            if (quantity <= 0)
            {
                continue;
            }

            if (market.PricesByItemId.TryGetValue(itemId, out int price))
            {
                total += quantity * price;
            }
            else if (data.ItemsById.TryGetValue(itemId, out ItemDefinition? item))
            {
                total += quantity * item.BasePrice;
            }
        }

        return total;
    }

    private static List<ItemDefinition> BuildAvailableItemsForPort(
        IReadOnlyList<ItemDefinition> allItems,
        PortZone zone,
        Random random)
    {
        List<ItemDefinition> zoneItems = allItems.Where(item => RarityForZone(zone) == item.Rarity).ToList();
        List<ItemDefinition> otherItems = allItems.Where(item => RarityForZone(zone) != item.Rarity).ToList();

        return PickDistinct(zoneItems, 4, random)
            .Concat(PickDistinct(otherItems, 2, random))
            .OrderBy(item => item.Name, StringComparer.Ordinal)
            .ToList();
    }

    private static IEnumerable<ItemDefinition> PickDistinct(List<ItemDefinition> candidates, int count, Random random)
    {
        List<ItemDefinition> remaining = new(candidates);
        List<ItemDefinition> selected = new();

        while (selected.Count < count && remaining.Count > 0)
        {
            int index = random.Next(remaining.Count);
            selected.Add(remaining[index]);
            remaining.RemoveAt(index);
        }

        return selected;
    }

    private static int CalculatePrice(ItemDefinition item, PortZone zone, Random random)
    {
        double variance = 0.3;
        double markup = GetItemMarkup(item.Rarity, zone);
        double multiplier = 1.0 + ((random.NextDouble() * 2.0 - 1.0) * variance) + markup;
        return Math.Max(1, (int)(item.BasePrice * multiplier));
    }

    private static ItemRarity RarityForZone(PortZone zone)
    {
        return zone switch
        {
            PortZone.Inner => ItemRarity.Common,
            PortZone.Outer => ItemRarity.MidTier,
            PortZone.Fringe => ItemRarity.Exotic,
            _ => ItemRarity.Common,
        };
    }

    private static double GetItemMarkup(ItemRarity rarity, PortZone zone)
    {
        return rarity switch
        {
            ItemRarity.Common when zone == PortZone.Fringe => 2.0,
            ItemRarity.Common when zone == PortZone.Outer => 0.5,
            ItemRarity.Common => 0.0,
            ItemRarity.MidTier when zone == PortZone.Fringe => 1.0,
            ItemRarity.MidTier when zone == PortZone.Inner => 0.5,
            ItemRarity.MidTier => 0.0,
            ItemRarity.Exotic when zone == PortZone.Inner => 2.0,
            ItemRarity.Exotic when zone == PortZone.Outer => 0.5,
            ItemRarity.Exotic => 0.0,
            _ => 0.0,
        };
    }
}

File: internal/services/economy_service.go
package services

import (
	"sort"

	"github.com/oberones/star-smuggler-go/internal/domain"
)

type EconomyService struct{}

func (s EconomyService) CreateInitialMarkets(data domain.DataSnapshot, rng RNG) map[string]domain.MarketSnapshot {
	result := make(map[string]domain.MarketSnapshot, len(data.Ports))
	for _, port := range data.Ports {
		result[port.ID] = s.CreateMarketForPort(port, data.Items, rng)
	}
	return result
}

func (s EconomyService) CreateMarketForPort(port domain.PortDefinition, items []domain.ItemDefinition, rng RNG) domain.MarketSnapshot {
	availableItems := buildAvailableItemsForPort(items, port.Zone, rng)
	priceMap := make(map[string]int, len(items))

	for _, item := range items {
		priceMap[item.ID] = calculatePrice(item, port.Zone, rng)
	}

	return domain.MarketSnapshot{
		PortID:           port.ID,
		AvailableItemIDs: availableItems,
		PricesByItemID:   priceMap,
	}
}

func (s EconomyService) RefreshAvailableGoods(run *domain.RunState, data domain.DataSnapshot, portID string, rng RNG) {
	port, ok := data.PortsByID[portID]
	if !ok {
		return
	}

	market, ok := run.MarketsByPortID[portID]
	if !ok {
		run.MarketsByPortID[portID] = s.CreateMarketForPort(port, data.Items, rng)
		return
	}

	market.AvailableItemIDs = buildAvailableItemsForPort(data.Items, port.Zone, rng)
	run.MarketsByPortID[portID] = market
}

func (s EconomyService) RefreshAllPrices(run *domain.RunState, data domain.DataSnapshot, rng RNG) {
	run.MarketsByPortID = s.CreateInitialMarkets(data, rng)
	run.JumpsSinceLastUpdate = 0
}

func (s EconomyService) GetCurrentMarket(run domain.RunState) (domain.MarketSnapshot, bool) {
	market, ok := run.MarketsByPortID[run.Player.CurrentPortID]
	return market, ok
}

func (s EconomyService) GetCargoLoad(run domain.RunState) int {
	return run.Cargo.TotalUnits()
}

func (s EconomyService) GetAvailableGoodsForCurrentPort(run domain.RunState, data domain.DataSnapshot) []domain.ItemDefinition {
	market, ok := s.GetCurrentMarket(run)
	if !ok {
		return nil
	}

	goods := make([]domain.ItemDefinition, 0, len(market.AvailableItemIDs))
	for _, itemID := range market.AvailableItemIDs {
		if item, exists := data.ItemsByID[itemID]; exists {
			goods = append(goods, item)
		}
	}
	return goods
}

func (s EconomyService) GetSellableCargoValueAtCurrentPort(run domain.RunState, data domain.DataSnapshot) int {
	market, ok := s.GetCurrentMarket(run)
	if !ok {
		return 0
	}

	total := 0
	for itemID, quantity := range run.Cargo.ItemQuantities {
		if quantity <= 0 {
			continue
		}

		if price, exists := market.PricesByItemID[itemID]; exists {
			total += quantity * price
			continue
		}

		if item, exists := data.ItemsByID[itemID]; exists {
			total += quantity * item.BasePrice
		}
	}

	return total
}

func buildAvailableItemsForPort(allItems []domain.ItemDefinition, zone domain.PortZone, rng RNG) []string {
	zoneRarity := rarityForZone(zone)
	zoneItems := make([]domain.ItemDefinition, 0)
	otherItems := make([]domain.ItemDefinition, 0)

	for _, item := range allItems {
		if item.Rarity == zoneRarity {
			zoneItems = append(zoneItems, item)
		} else {
			otherItems = append(otherItems, item)
		}
	}

	selected := append(pickDistinct(zoneItems, 4, rng), pickDistinct(otherItems, 2, rng)...)
	sort.Slice(selected, func(i, j int) bool {
		return selected[i].Name < selected[j].Name
	})

	ids := make([]string, 0, len(selected))
	for _, item := range selected {
		ids = append(ids, item.ID)
	}
	return ids
}

func pickDistinct(candidates []domain.ItemDefinition, count int, rng RNG) []domain.ItemDefinition {
	remaining := append([]domain.ItemDefinition(nil), candidates...)
	selected := make([]domain.ItemDefinition, 0, count)

	for len(selected) < count && len(remaining) > 0 {
		index := rng.Intn(len(remaining))
		selected = append(selected, remaining[index])
		remaining = append(remaining[:index], remaining[index+1:]...)
	}

	return selected
}

func calculatePrice(item domain.ItemDefinition, zone domain.PortZone, rng RNG) int {
	const variance = 0.3
	markup := getItemMarkup(item.Rarity, zone)
	multiplier := 1.0 + ((rng.Float64()*2.0 - 1.0) * variance) + markup
	price := int(float64(item.BasePrice) * multiplier)
	if price < 1 {
		return 1
	}
	return price
}

func rarityForZone(zone domain.PortZone) domain.ItemRarity {
	switch zone {
	case domain.PortZoneInner:
		return domain.ItemRarityCommon
	case domain.PortZoneOuter:
		return domain.ItemRarityMidTier
	case domain.PortZoneFringe:
		return domain.ItemRarityExotic
	default:
		return domain.ItemRarityCommon
	}
}

func getItemMarkup(rarity domain.ItemRarity, zone domain.PortZone) float64 {
	switch {
	case rarity == domain.ItemRarityCommon && zone == domain.PortZoneFringe:
		return 2.0
	case rarity == domain.ItemRarityCommon && zone == domain.PortZoneOuter:
		return 0.5
	case rarity == domain.ItemRarityMidTier && zone == domain.PortZoneFringe:
		return 1.0
	case rarity == domain.ItemRarityMidTier && zone == domain.PortZoneInner:
		return 0.5
	case rarity == domain.ItemRarityExotic && zone == domain.PortZoneInner:
		return 2.0
	case rarity == domain.ItemRarityExotic && zone == domain.PortZoneOuter:
		return 0.5
	default:
		return 0.0
	}
}
